Initial composite/global Propagators update for OTEP 66. (#440)
open-telemetry/opentelemetry-specification: specification/api-propagators.md

  - [Extract](#extract)
    - [Getter argument](#getter)
      - [Get](#get)
- [Composite Propagator](#composite-propagator)
- [Global Propagators](#global-propagators)

</details>

> ## Overview

Reviewer: It may be worth introducing the notion of "application message" that semantically consists of metadata (e.g. HTTP headers) and the payload. The propagators' job is to encode/decode a representation of context state for a given cross-cutting concern to/from the message metadata.

Cross-cutting concerns send their state to the next process using
`Propagator`s, which are defined as objects used to read and write

> context data to and from RPC requests. Each concern creates a set

Author: The original intent was to include all remote calls (which would include HTTP requests and the sort). Wondering what to use instead of this, to not actually list HTTP and other systems ;)

> context data to and from RPC requests. Each concern creates a set

Reviewer 2: What about "_to write context data to requests being sent and read it from requests being received_"?

> context data to and from RPC requests. Each concern creates a set

Reviewer: replace "requests" with "messages exchanged by the applications"

> of `Propagator`s for every supported format.

Reviewer: "format" should be styled as a keyword.

## HTTP Text Format

`HTTPTextFormat` is a formatter that injects and extracts a cross-cutting concern

> value as text into carriers that travel in-band across process boundaries.

Reviewer: * I would avoid using the term "formatter"
* This sentence seems to repeat the definition of the Propagator. Whereas "Format" is a representation of _restrictions_ that a given message transport imposes on the metadata, such as:
  * Metadata is a collection of string->string pairs (as opposed to Binary format) - this implies a specific Carrier API
  * HTTP header keys are case insensitive
  * Comma is not allowed as HTTP header value
  * etc.

Ideally we would decouple the Carrier API from the format's syntax restrictions, but it's pretty complicated to do.

Reviewer: NB: such explanation should precede the section on HTTP Text Format, since it is a generic description of the Propagator role, with HTTP just being an example.

> 

Reviewer: I do not follow anything in the preceding Fields section. It seems very vague and random. And certainly does not belong specifically under the HTTP format.

The Get function is responsible for handling case sensitivity. If the getter is intended to work with a HTTP request object, the getter MUST be case insensitive. To improve compatibility with other text-based protocols, text format implemenations MUST ensure to always use the canonical casing for their attributes. NOTE: Cannonical casing for HTTP headers is usually title case (e.g. `Content-Type` instead of `content-type`).

## Composite Propagator

Implementations MUST offer a facility to group multiple `Propagator`s
from different cross-cutting concerns in order to leverage them as a
single entity.

The resulting composite `Propagator` will invoke the `Propagators`
in the order they were specified.

Each composite `Propagator` will be bound to a specific format, such
as `HttpTextFormat`, as different formats will likely operate on different
data types.
There MUST be functions to accomplish the following operations.

- Create a composite propagator
- Extract from a composite propagator
- Inject into a composite propagator

### Create a Composite Propagator

Required arguments:

- A list of `Propagator`s.

Returns a new composite `Propagator` with the specified `Propagator`s.

> ### Extract

Reviewer 2: Huh, are the headings _Extract_ and _Inject_ mixed up or am I missing something?
I'd expect the Extract operation to take a Getter and Inject to take a Setter.

Author: Oops, thanks, fixed ;)

Required arguments:

- A `Context`.
- The carrier that holds propagation fields.

> - the `Setter` invoked for each propagation key to add or remove.

Reviewer 2: ```suggestion
- The `Setter` invoked for each propagation key to add or remove.
```
consistency 💪

> ### Inject

Reviewer 3: Are we going to specify whether inject mutates the carrier or returns a new one?

Author: So currently the `carrier` that is passed is in turn passed to the `Setter` instance, which decides what to do with it along with the keys/values. In practice (99% of the time?) the carrier is indeed mutated, but not 'enforced'.

Reviewer 3: I was going to mark this as resolved, but for some reason don't have the button available like in the spec repo. Are permissions different here?

Reviewer: The Carrier APIs should be defined earlier in the doc. They are generally mutable by definition, although I could see how this could irk some FP-oriented people. It is rather difficult to make individual pieces immutable and composable, because the Carrier API fundamentally implies some transformation that is unknown to the propagator, like writing every key-value pair as HTTP header is a specific implementation of HTTPRequest.

Reviewer: (well, it's actually not THAT difficult, using some pipeline like `propagators.map(_.stateAsMap(context)).flatMap(carrier.accumulate)`, 
but it's rather inefficient).

Reviewer 2: @dyladan You need to be the author who opened the PR or have write permissions to the repo (i.e., be in spec-approvers) to resolve conversations.

Required arguments:

- A `Context`.
- The carrier that holds propagation fields.

> - The `Setter` invoked for each propagation key to add or remove.

Reviewer: Why do we need to separate carrier and setter?

## Global Propagators

> Implementations can optionally provide global `Propagator`s for

Reviewer 2: ```suggestion
Implementations MAY provide global `Propagator`s for
```

Author: Wondering if this will cause slight confusion in the future, but lets go with it and put "optionally" back if issues arise.

each supported format.

If offered, the global `Propagator`s should default to a composite `Propagator`

> containing W3C TraceContext and CorrelationContext `Propagator`s,

Reviewer 2: ```suggestion
containing W3C Trace Context and Correlation Context `Propagator`s,
```
nit: spelled like this in the W3C specs

in order to provide propagation even in the presence of no-op

> OpenTelemetry implementations.

Reviewer 2: There was no agreement reached on #208 (#428) regarding that yet, right?

Author: Before OTEP 66 this was already stated (see `api-tracing.md` for example). Once #208 is resolved, but might need to update this one though :)

> ### Get Global Propagator

Reviewer 3: Do we want one global propagator (which may be composite), or do we want to add the propagator to the TracerProvider? Reason being that if you have multiple apps in a single runtime, each of which has its own tracer provider, you may want to use different propagation formats between apps.

Author: To me doing that sounds like an overkill _for now_ - in practice, it looks like most people want a single place/component to do the entire injection/extraction.

We could add in the future an additional `Propagator` component to items such as `TracerProvider` and similar components (defaulting to `null`) in case users want to (optionally) override the global one, though 😃

Reviewer 3: How should we be calling the propagator?

1. `api.propagator.extract(...)`
2. `tracer.extract(...)`
3. `tracer_provider.extract(...)`

Author: You should be using 1) as this will be the 'default' thing - we might (but we might not) add 3) as a custom, override thing, in which case we could call that one as well.

The second option is due to be removed ;)

- the key of the field.

The Get function is responsible for handling case sensitivity. If the getter is intended to work with a HTTP request object, the getter MUST be case insensitive. To improve compatibility with other text-based protocols, text format implemenations MUST ensure to always use the canonical casing for their attributes. NOTE: Cannonical casing for HTTP headers is usually title case (e.g. `Content-Type` instead of `content-type`).

## Composite Propagator

Implementations MUST offer a facility to group multiple `Propagator`s
from different cross-cutting concerns in order to leverage them as a
single entity.

The resulting composite `Propagator` will invoke the `Propagators`
in the order they were specified.

Each composite `Propagator` will be bound to a specific format, such

> as `HttpTextFormat`, as different formats will likely work on a different

Reviewer 4: Is this last clause trying to say that formats are not intercompatible?

We might want to rephrase if that is the intent.

Author: That's correct, propagators for different formats are not inter compatible.

Reviewer 4: can we rephrase to:

```suggestion
as `HttpTextFormat`, as different formats will likely operate on different
```

I'm worried that `work` will be interpreted as being compatible.

Author: Lets do that.

data types.

> 

Reviewer 4: The flow from this section to the subsections is rather abrupt. Can we outline the next sections as operations that need to be supported?

```suggestion
There MUST be functions to accomplish the following operations.

- Create a composite propagator
- Extract from a composite propagator
- Inject into a composite propagator

```

### Create a Composite Propagator

Required arguments:

- A list of `Propagator`s.

Returns a new composite `Propagator` with the specified `Propagator`s.

### Extract

> - A `Context`.

Reviewer 4: ```suggestion
Required arguments:

- A `Context`.
```

- The carrier that holds propagation fields.

> - the `Setter` invoked for each propagation key to add or remove.

Reviewer 4: Does the `Setter` act on the carrier passed as the second argument?

Author: So this will act exactly as the `Propagator` for `HttpTextFormat`, so in theory there's no need to repeat the explanation here (once #424 is merged). We will need a very clear line in the Overview section mentioning this fact, of course ;)

> 

Reviewer 4: Should this operation return something?

Author: Yes, it does - that's covered in #424 though ;)

### Inject

> - A `Context`.

Reviewer 4: ```suggestion
Required arguments:

- A `Context`.
```

- The carrier that holds propagation fields.
- The instance of `Getter` invoked for each propagation key to get.

## Global Propagators

Implementations can optionally provide global `Propagator`s for
each supported format.

If offered, the global `Propagator`s should default to a composite `Propagator`
containing W3C TraceContext and CorrelationContext `Propagator`s,
in order to provide propagation even in the presence of no-op

> OpenTelemetry implementations.

Reviewer 5: :+1:

> ### Get Propagator

Reviewer 5: Should this section read "Get Global Propagator"?

This method MUST exist for each supported format.

Returns a global `Propagator`. This usually will be composite instance.

> ### Set Propagator

Reviewer 5: "Set Global Propagator"?
